platform/viewer/src/connectedComponents/ToolbarRow.js

```diff
@@ -279,21 +267,38 @@ function _getVisibleToolbarButtons() {
   return toolbarButtonDefinitions;
 }
 
-function _handleBuiltIn({ behavior } = {}) {
+function _handleBuiltIn({ id, options: { behavior } } = {}) {
+  /* TODO: Keep cine button active until its unselected. */
+  const { dialog, modal, t } = this.props;
+  const { dialogId } = this.state;
+
   if (behavior === 'CINE') {
-    this.setState({
-      isCineDialogOpen: !this.state.isCineDialogOpen,
-    });
+    if (dialogId) {
+      dialog.dismiss({ id: dialogId });
+      this.setState({ dialogId: null });
+    } else {
+      const spacing = 20;
+      const { x, y } = document
+        .querySelector(`.ViewerMain`)
+        .getBoundingClientRect();
+      const newDialogId = dialog.create({
+        content: ConnectedCineDialog,
+        defaultPosition: {
+          x: x + spacing || 0,
+          y: y + spacing || 0,
+        },
+      });
+      this.setState({ dialogId: newDialogId });
+    }
```

Comment:
This is great. Here is something we need to think about:

How can we make extensions powerful enough that this logic can live in them? Having "Built in" viewer behavior lets us offload some concerns to `viewer`, but it also makes us less flexible.

Ideally, we end up at something like:

- ToolbarButton is of type `Toggle`
  - When clicked, is `enabled`, and extension's logic is called w/ `enabled: true`
  - When clicked again, enabled is `false`, and extension's logic is called
- The toolbar can use the now exposed `Dialog` service to hide/show dialog
- **As the user interacts, the extension is able to update the ViewportComponent based on that interaction**

I think that last part is the only thing missing, and what we need a good solution for (outside the scope of this PR, but an upcoming issue for sure). If we wanted to update... cineIsPlaying, stack index, slice index, etc. What would that look like for an extension?

I'll create a `discussion` issue for this.

Reply:
Today, we only have two viewport extensions that can leverage CINE like behavior. I could see that growing AND shrinking. I am a little torn on if this **specific** use case makes more sense to configure in `@ohif/viewer` or `@ohif/extension-*`. Our current implementation updates viewport props, and it's up to registered viewports to leverage those prop updates to show CINE behavior.

The changes outlined above would make it easier for extensions to:

1. Show custom dialog/modal UI
2. Handle how input to that UI influences behavior _like_ viewport CINE behavior

But this could also extend to modifying any state the `@ohif/viewer` has traditionally managed so far. One thing we have discussed is creating a `MeasurementsService`.

Reply:
Follow-up issue created here: https://github.com/OHIF/Viewers/issues/1205